src/actions/commands/fold.ts

```diff
@@ -72,3 +75,31 @@ class CommandOpenAllFoldsRecursively extends CommandFold {
   keys = ['z', 'O'];
   commandName = 'editor.unfoldRecursively';
 }
+
+abstract class FoldOperator extends BaseOperator {
+  public modes = [Mode.Normal];
+  abstract commandName: string;
+
+  public async run(vimState: VimState, start: Position, end: Position): Promise<void> {
+    const previousSelections = vimState.editor.selections;
+    vimState.editor.selection = new vscode.Selection(start, end);
+    await vscode.commands.executeCommand(this.commandName);
+    await new CommandCloseFold().exec(start, vimState);
+    vimState.editor.selections = previousSelections;
```

Comment:
Why set `vimState.editor.selections` then set `vimState.cursors`?

The latter should generally be the source of truth and `ModeHandler` will bring the former into line with it after the action's been run. Setting `vimState.editor.selections` directly in an action is sometimes necessary (for stupid reasons) but should be avoided because it can trigger `TextEditorSelectionChangeEvent` callbacks, which can cause issues.

Reply:
So, the previous selections are relevant in `'<'>` commands, so I have to save and restore them.
The cursor position is the position of where the cursor should be set after the folding, matching Vim's behavior.

Reply:
`'<'>` relies on `vimState.lastVisualSelection` - can't we just make sure that's preserved?

Reply:
Possibly. I'll look into it 👍.

Reply:
Okay. Investigating it also led me to find few bugs. Seems to be corrected now. I'll see if I can add some tests (is testing folds is even possible?).

Reply:
> is testing folds is even possible?

Probably not easily - I wouldn't worry about it too much for now